src/util/records.py

```diff
@@ -1250,9 +1253,46 @@ def delete_unused(cr, *xmlids, **kwargs):
         table = table_of_model(cr, model)
         res_id_to_xmlid = dict(zip(ids, xids))
 
-        sub = " UNION ".join(
+        cascade_children = [
+            fk_col
+            for fk_tbl, fk_col, _, fk_act in get_fk(cr, table, quote_ident=False)
+            if fk_tbl == table and fk_act == "c"
+        ]
+        if cascade_children:
+            if len(cascade_children) == 1:
+                join = format_query(cr, "t.{}", cascade_children[0])
+            else:
+                join = (
+                    sql.SQL("ANY(ARRAY[")
+                    + sql.SQL(", ").join(sql.Composed([sql.SQL("t."), sql.Identifier(cc)]) for cc in cascade_children)
+                    + sql.SQL("])")
+                )
+
+            kids_query = format_query(
+                cr,
+                """
+                WITH RECURSIVE _child AS (
+                    SELECT id AS root, id AS child
+                      FROM {0}
+                     WHERE id = ANY(%(ids)s)
+                     UNION  -- don't use UNION ALL in case we have a loop
+                    SELECT c.root, t.id as child
+                      FROM {0} t
+                      JOIN _child c
+                        ON c.child = {1}
+                )
+                SELECT root, child
+                  FROM _child
+                """,
+                table,
+                join,
+            )
+        else:
+            kids_query = format_query(cr, "SELECT id AS root, id AS child FROM {0} WHERE id = ANY(%(ids)s)", table)
+
+        sub = " UNION ALL ".join(
```

Comment:
maybe we can use SQLStr
```suggestion
        sub = SQLStr(" UNION ALL ".join(
```